### ISSUE TITLE
Failure to downsize nodepools

### ISSUE THREAD

**Issue by ido-opster** [Opster/opensearch-k8s-operator#227]
Let's say cluster is up and running with 1 master node.
Now i want to increase cluster master nodes to 3 so i set crd node pools replica:3
Operator will try to add one master node at a time. 
Operator will stay stuck for always because second master node will never be ready due to split brain issue.

**Comment by swoehrl-mw** [comment]
As we officially not (yet) support single node clusters I think this is more an enhancement request than a bug. 
And personally I think this is an edge case that we should not support as IMO we also should not support single-node clusters.

**Comment by dickescheid** [comment]
I just tried increasing from 3 to 4 masters, and it worked.

```
2022-08-19T08:42:07,177][INFO ][o.o.c.s.ClusterApplierService] [testos2-masters-3] master node changed {previous [], current [{testos2-masters-0}{XG6XWKeBQ76eCKVXcu9r4w}{NTmNue6dTsGJuiogVsDcHw}{testos2-masters-0}{10.92.4.209:9300}{dm}{shard_indexing_pressure_enabled=true}]}, added {{testos2-masters-2}{H7E9eVHWQHGIQyIPahCxfA}{QTZ0QJyuSCm1OlNuQR3uXQ}{testos2-masters-2}{10.92.0.69:9300}{dm}{shard_indexing_pressure_enabled=true},{testos2-masters-1}{tGEMPaRkSbG7VVnGdy22SQ}{1_ZtUToGTXyYJ1jmMc2Ihw}{testos2-masters-1}{10.92.5.26:9300}{dm}{shard_indexing_pressure_enabled=true},{testos2-masters-0}{XG6XWKeBQ76eCKVXcu9r4w}{NTmNue6dTsGJuiogVsDcHw}{testos2-masters-0}{10.92.4.209:9300}{dm}{shard_indexing_pressure_enabled=true}}, term: 3, version: 42, reason: ApplyCommitRequest{term=3, version=42, sourceNode={testos2-masters-0}{XG6XWKeBQ76eCKVXcu9r4w}{NTmNue6dTsGJuiogVsDcHw}{testos2-masters-0}{10.92.4.209:9300}{dm}{shard_indexing_pressure_enabled=true}}
[2022-08-19T08:42:07,279][INFO ][o.o.c.s.ClusterSettings  ] [testos2-masters-3] updating [plugins.index_state_management.metadata_migration.status] from [0] to [1]
[2022-08-19T08:42:07,279][INFO ][o.o.c.s.ClusterSettings  ] [testos2-masters-3] updating [plugins.index_state_management.template_migration.control] from [0] to [-1]
[2022-08-19T08:42:07,372][INFO ][o.o.a.c.HashRing         ] [testos2-masters-3] Node added: [yQCCGJQoRbyhFUQpezoqPg, tGEMPaRkSbG7VVnGdy22SQ, H7E9eVHWQHGIQyIPahCxfA, XG6XWKeBQ76eCKVXcu9r4w]
[2022-08-19T08:42:07,378][INFO ][o.o.a.c.ADClusterEventListener] [testos2-masters-3] Cluster node changed, node removed: false, node added: true
[2022-08-19T08:42:07,379][INFO ][o.o.a.c.HashRing         ] [testos2-masters-3] AD version hash ring change is in progress. Can't build hash ring for node delta event.
[2022-08-19T08:42:07,379][INFO ][o.o.a.c.ADClusterEventListener] [testos2-masters-3] Hash ring build result: false
[2022-08-19T08:42:07,573][INFO ][o.o.h.AbstractHttpServerTransport] [testos2-masters-3] publish_address {testos2-masters-3/10.92.1.4:9200}, bound_addresses {0.0.0.0:9200}
[2022-08-19T08:42:07,573][INFO ][o.o.n.Node               ] [testos2-masters-3] started
```

**Comment by dickescheid** [comment]
inversely, can it be possible to decrease the number of nodes? 
I tried applying a decreased number of nodes, but the statefulset is not updated, thus kubernetes will not reduce the number of nodes.

**Comment by swoehrl-mw** [comment]
Hi @dickescheid:
> I tried applying a decreased number of nodes, but the statefulset is not updated, thus kubernetes will not reduce the number of nodes.

Could you please check the logs of the operator itself for any reported errors when doing the scale down?
I tried to recreate your problem by creating a 4 node cluster and then resizing to 3 and got an error about `failed to create os client`. it would be interesting to see if that is only my local test setup or if that could be the source of your problem also.

**Comment by swoehrl-mw** [comment]
Thanks for testing @dickescheid. This proves my theory. 
When downsizing the cluster the operator first drains and excludes the opensearch node about to be removed. For this it uses the OpenSearch REST API. Currently the code is a bit of a mix, some calls connect to opensearch using the service cluster dns name, while others create a nodeport service and connect through that. And these calls fail. I'll open a PR to move all calls to use cluster dns, that should fix the problem.

**Comment by dickescheid** [comment]
nice, thanks!

**Comment by swoehrl-mw** [comment]
Should be fixed with 2.1.0. Closing as completed.
